wave_matching/tests/icp_tests.cpp

```diff
@@ -0,0 +1,111 @@
+
+#include "wave/matching/ICP.hpp"
+
+namespace wave {
+
+bool isApprox(const Eigen::Affine3d &first,
+              const Eigen::Affine3d &second,
+              double tolerance) {
+    for (int i = 0; i < 4; ++i) {
+        for (int j = 0; j < 4; ++j) {
+            tolerance -= std::abs(first(i, j) - second(i, j));
+        }
+    }
+    if (tolerance > 0) {
+        return true;
+    }
+    return false;
+}
+
+// Fixture to load same pointcloud all the time
+class ICPTest : public testing::Test {
+ protected:
+    ICPTest() {}
+
+    virtual ~ICPTest() {
+        if (matcher) {
+            delete matcher;
+        }
+    }
+
+    virtual void SetUp() {
+        this->ref = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+        this->target = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+        pcl::io::loadPCDFile("../../wave_matching/tests/testscan.pcd",
+                             *(this->ref));
+    }
+
+    void setParams(const float res, const Eigen::Affine3d perturb) {
+        this->matcher = new ICP_Matcher(res);
+        pcl::transformPointCloud(*(this->ref), *(this->target), perturb);
+        this->matcher->setup(this->ref, this->target);
+    }
+
+    pcl::PointCloud<pcl::PointXYZ>::Ptr ref, target;
+    ICP_Matcher *matcher;
+};
+
+TEST(ICPTests, initialization) {
+    ICP_Matcher matcher(0.1f);
+}
+
+// Zero displacement without downsampling
+TEST_F(ICPTest, fullResNullMatch) {
+    Eigen::Affine3d perturb;
+    Eigen::Affine3d result;
+    bool match_success = false;
+    // Setup
+    perturb = Eigen::Affine3d::Identity();
+    perturb.translation() << 0, 0, 0;
+    setParams(-1, perturb);
+    // test and assert
+    match_success = matcher->match();
+    result = matcher->getResult();
+    EXPECT_TRUE(match_success);
+    EXPECT_TRUE(isApprox(result, perturb, 0.1));
+}
+
+// Zero displacement using voxel downsampling
+TEST_F(ICPTest, nullDisplacement) {
+    Eigen::Affine3d perturb;
+    Eigen::Affine3d result;
+    bool match_success = false;
+    // Setup
+    perturb = Eigen::Affine3d::Identity();
+    perturb.translation() << 0, 0, 0;
+    setParams(0.05f, perturb);
+    // test and assert
+    match_success = matcher->match();
+    result = matcher->getResult();
+    EXPECT_TRUE(match_success);
+    EXPECT_TRUE(isApprox(result, perturb, 0.1));
+}
+
+// Small displacement using voxel downsampling
+TEST_F(ICPTest, smallDisplacement) {
+    Eigen::Affine3d perturb;
+    Eigen::Affine3d result;
+    bool match_success = false;
+    // Setup
+    perturb = Eigen::Affine3d::Identity();
+    perturb.translation() << 0.2, 0, 0;
+    setParams(0.05f, perturb);
+    // test and assert
+    match_success = matcher->match();
+    result = matcher->getResult();
+    EXPECT_TRUE(match_success);
+    EXPECT_TRUE(isApprox(result, perturb, 0.1));
+}
+
+}  // end of namespace wave
+
+int main(int argc, char *argv[]) {
+    testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
```

Comment:
At current I have a general wave_utils_tests and wave_optimization_tests test runner for each module of libwave, maybe you should do the same? For example this is how I link up all my test cases into a single test runner in the utils module:

```
# UNIT TESTS
SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/tests)
FILE(COPY tests/data DESTINATION ${PROJECT_BINARY_DIR}/tests)

ADD_EXECUTABLE(
    # TARGET
    wave_utils_tests
    # TESTS
    tests/utils/config_test.cpp
    tests/utils/data_test.cpp
    tests/utils/filesystem_test.cpp
    tests/utils/math_test.cpp
    tests/utils/time_test.cpp
    # TEST RUNNER
    tests/test_runner.cpp
)
TARGET_LINK_LIBRARIES(wave_utils_tests ${TEST_DEPS})
```
Open to suggestions if you have alternatives.